### PR TITLE
Z21 pending request queue

### PR DESCRIPTION
**Pull request by gfgit**
Pending request queue allows to retry sending messages after a timeout if expected reply is not received.
This also allows to detect a message is a reply to our own changes and react to it differently

See #60 for earlier discussion.

NOTE: needs #82 applied first

Both this and #82 `LocoCache` try to resolve the "async loop" issue (See #57).
This 2 solutions are orthogonal and can live together.

### CODE REVIEW

**Comment by gfgit** [comment]
Quick comparison with #59 approach (not fully accurate):

## **LocoCache** approach
PROS:

- Less memory footprint when number of loco is low and many messages are sent
- Can extract individual changes from `LanXLocoInfo` like change only direction or only emergency stop
  This is possible because we cache last value of direction, speed and emergency stop per each loco.
- Can somewhat detect speed trend change: user reduces speed while train is accelerating or viceversa

CONS:

- Weak logic based on heuristic. Many edge cases
- Logic is more obscure
- More memory when many loco are running
- Annoying "ignore" delay when user moves throttle right after Traintastic has updated same loco
  This is needed to avoid receiving our own reply but creates discrepancy between real loco state and cached one.

## Pending Request Queue approach

PROS:
- While logic is more complex, it's easier to understand
-  More accurate for own replies (which get then ignored)
- Expandable to other message types
- Allows re-sending a message if no reply is received so it workarounds UDP packet loss
- Logic is much more robust
- No delay needed for throttle messages

CONS:
- Can detect external changes as our own replies if messages match.
  This shouldn't be a big problem but there are at least some edge cases (see below)
- Potential risk of big memory footprint is sending a lot of messages in a fast manner
- Does not check received message order so 2 replies referring to same object can be
  detected in wrong order.
- Cannot extract individual changes because it has no track of last value

## Future work

The solution of course will be "best of both worlds". I'll try to tune both approaches to work together
and understand which one is doing what exactly and what are the edge cases left.

It already shows big improvement when simulating enormous network delays (100 ms for every message on Z21 side)
but it's not yet perfect.
It really needs testing with real hardware (which unfortunately I don't have) because different behaviors might arise

**Comment by reinder** [comment]
Looking good, I really like it. Nice work! I don't own a z21 or Z21 yet so can't test it against a real one, I only have a Digikeijs DR5000 which supports the Z21 protocol.

> If you get speed 15 after speed 5 it means someone else has explicitly set speed to 15

Or that 6 .. 14 are dropped, it's UDP after all. But under normal circumstances without a high network load it probably won't happen.

Maybe you can add a debug option to log some additional info about the retries, that might be useful if we find people to do some testing with it.

p.s. As the Z21 implementation is receiving many improvements, it seems fair to me that you add a Copyright line for your name. (Don't know how this normally works...this is my first Open Source project with contributions :))

**Comment by gfgit** [comment]
Rebased onto #82

File: server/src/hardware/protocol/z21/messages.hpp [hunk 1]
```diff
@@ -1423,6 +1444,69 @@ constexpr std::string_view toString(LanXSetLocoFunction::SwitchType value)
...
 }
 
+struct MessageReplyType
+{
+  enum Priority
+  {
+    LowPriority = 0,
+    Normal = 1,
+    Urgent = 2
+  };
```
Comment (reinder) [inline server/src/hardware/protocol/z21/messages.hpp:1454]: Why does one element have `Priority` in it and the other don't have it?

Change it to `enum class Priority : uint8_t`? Then the values are scoped :)

File: server/src/hardware/protocol/z21/messages.hpp [hunk 2]
```diff
@@ -1423,6 +1444,69 @@ constexpr std::string_view toString(LanXSetLocoFunction::SwitchType value)
...
+
+  enum Flags
```
Comment (reinder) [inline server/src/hardware/protocol/z21/messages.hpp:1456]: `enum class`?

File: server/src/hardware/protocol/z21/messages.hpp [hunk 3]
```diff
@@ -1423,6 +1444,69 @@ constexpr std::string_view toString(LanXSetLocoFunction::SwitchType value)
...
+  {
+    CheckDb0       = 1,
+    CheckSpeedStep = 1 << 1,
+    CheckAddress   = 1 << 2
+  };
+
+  inline Priority getPriority() const
```
Comment (reinder) [inline server/src/hardware/protocol/z21/messages.hpp:1463]: remove `get` for getter, the Traintastic codebase style doesn't do that.

File: server/src/hardware/protocol/z21/messages.hpp [hunk 4]
```diff
@@ -1423,6 +1444,69 @@ constexpr std::string_view toString(LanXSetLocoFunction::SwitchType value)
...
+  {
+    return Priority(m_flags & 0xF);
+  }
+
+  inline void setPriority(Priority priority)
```
Comment (reinder) [inline server/src/hardware/protocol/z21/messages.hpp:1468]: as the above is change, the argument name causes a *shadow* warning, that's why the common practice in the Traintastic codebase is to use `value` as name for the setter argument.

**Review by reinder — CHANGES_REQUESTED** [review]

File: server/src/hardware/protocol/z21/messages.cpp [hunk 5]
```diff
@@ -324,4 +324,187 @@ bool LanX::isChecksumValid(const LanX &lanX)
   return XpressNet::isChecksumValid(msg, dataSize);
 }
 
+MessageReplyType getReplyType(const Message &message)
+{
+  switch (message.header())
```
[the remark this answers is not in the document]
Comment (reinder) [reply server/src/hardware/protocol/z21/messages.cpp:330]: yeah, clang-format would really help, I made one some time ago, that was working quiet well. There were a few issues, please remember me again if I forget :)

**Comment by gfgit** [comment]
I now have built a Z21 with ESP32 and did some initial tests. Pending reply queue is the way to go.
We will keep some bits of my old `LocoCache` and remove it's timed logic and speed trend logic which never worked right and are difficult to understand.
Also `LocoCache` should get a better name.
I will clean the code and do some more tests

**Comment by gfgit** [comment]
Last commit is a fix unrelated to this PR but found when testing.
Also currently you cannot really test this PR in simulation mode because IO handler will not reply to locomotive drive messages, this is detected as UDP packet drop so messages are sent again 2 times.

**Comment by gfgit** [comment]
@reinder What do you think about this?

**Comment by reinder** [comment]
> Also currently you cannot really test this PR in simulation mode because IO handler will not reply to locomotive drive messages, this is detected as UDP packet drop so messages are sent again 2 times.

Can we add that in the simulation?

**Comment by gfgit** [comment]
> > Also currently you cannot really test this PR in simulation mode because IO handler will not reply to locomotive drive messages, this is detected as UDP packet drop so messages are sent again 2 times.
> 
> Can we add that in the simulation?

Yes we could but since we do not cache decoder state in simulation IO handler, the reply would contain incorrect values.

**Comment by gfgit** [comment]
The problem is that reply is always a `LAN_X_LOCO_INFO` which contains all loco state. So if we rely to decoder throttle change we need to put in also functions state which we do not know because we do not cache.
By putting them all to off we would break behavior of throttle widget for example because function would stay on only until a new reply is received.
If you want I can make something similar to LocoChache inside simulation IO handler to store decoder state for replies

**Comment by gfgit** [comment]
Proof of concept. Not tested yet

**Comment by reinder** [comment]
Looking good, it is really helpful to have proper simulation, this makes it easier to do development without hardware attached :) (We still need to test  with real hardware off course)

**Review by reinder — APPROVED** [review]

**Comment by reinder** [comment]
@gfgit Just tested simulation mode, works nicely. Anything else we need to check? (Or can I merge it?)

**Comment by gfgit** [comment]
You can merge it!

**Pull request merged**